futures-util/src/io/mod.rs

```diff
@@ -500,6 +505,61 @@ pub trait AsyncWriteExt: AsyncWrite {
         WriteAll::new(self, buf)
     }
 
+    /// Write data into this object.
+    ///
+    /// Creates a future that will write the entire contents of `bufs` into this
+    /// `AsyncWrite` using [vectored writes].
+    ///
+    /// The returned future will not complete until all the data has been
+    /// written.
+    ///
+    /// [vectored writes]: std::io::Write::write_vectored
+    ///
+    /// # Notes
+    ///
+    /// Different to `io::Write::write_vectored` this takes a *mutable*
+    /// reference to a slice of `IoSlice`s, not a non-mutable reference, because
+    /// we need to modify the slice to keep track of the bytes already written.
```

Comment:
Who needs to? The future can store that state internally. Is this for the user to know? IMO we should try to match the API of `write_all` as closely as possible. Actually, for that matter, what we should be trying to match would be the yet-to-exist `std::io::Write::write_all_vectored`. If there isn't yet an issue for creating that API, would you mind creating one and linking it here? I'd really like to keep these extension traits from diverging from the API in the standard library, but I'm fine landing it here if it's already under a feature gate, so long as we make sure to tag that feature gate as `_unstable` / have it require an `unstable` feature flag in order to compile. Does that make sense?

Reply:
The problem with writing all buffers is that is we need to deal with partially written `IoSlice`s. For which I wrote [`IoSlice::advance`](https://doc.rust-lang.org/std/io/struct.IoSlice.html#method.advance), which internally updates the slices, but needs mutable access. Basically `WriteAllVectored` needs ownership of `bufs`, to keep track of what bytes are already written. Also see the next section in the documentation.

In practice I don't think this will be a problem as `IoSlice` *slices*  are often created just before writing, e.g. see the example. But if you know of a better solution I'll love to hear it.

Reply:
Created https://github.com/rust-lang/rust/issues/70436.